### DISCUSSION TITLE
Forma Requirements 2023

### DISCUSSION THREAD

**Discussion by michaem** [formatools/forma#114] [General]
_Here will be description of all requirements for modern Android development with scalable structure and project architecture._
_Forma should help build project architecture with stable scalable metrics: build time, weak deps, feature extending for 50-100+ mobile engineers per project._

Killer metrics of Forma are STP:
- Providing **scalable** to your project
- Carefully to your **build time**
- Boost your **performing** and TTM

**Comment by michaem** [comment]
### Build System requirements

**1. Used easy setup project structure from `settings.gradle`**
_Prefered auto-generatation from existed module structure already_

**2. Provided dependencies management and them versions control**
_Now deps provided from buildSrc paradigm and versionsCatalog features_

**3. SImple plugging Forma tools for outer projects**
_Just describe your artifactory/nexus/maven repo and directly pass credentions to them, if necessary_

**4. Doesn't use `buildSrc ` paradigm**
_Forma themself provides deps, versions, changes to all project-subprojects and carefully for cache_

**5. Shortly and clearly plugged other root pluggins**
_You can easy plug root pluggins sush as: kotlin-stdlib, agp, gradle pluggin and etc. They shared for all subproject_

**6. Easy applied pluggins for subprojects**
_Just apply necessary plugging to your module and you're ready (may be some concern with module semantic validation)_

**7. Easy plugging dependencies for extraPlugins**
_Forma provided by using simply context '.' plugging any deps from deps-source (p. 2), if it doens't broke module type semantinc restrictions_

**8. Provided Build Type management**
_Just describe, what're build types project needs (debug, release, custom) and you'll have it with shrinking rules_

**9. Scalable multi-module structure**
_You don't need monolith modules types, such as: androidLibrary. Just using type-safe module types: api, impl, lib, util, you can build scalable multi-module structure, with carefuly for caching, all-ready deps rules between modules (api to api, but impl not to other impl)_ and build features (compose, androidRes) underhood.

**10. Provided declarative building configs**
_Ofen you need to pass some configs into 3thd part systems, such as analytics for subproject_

**11. Separated plugging for Forma features**
_You choose only Forma features, that necessary by plugging them separatly standalone_

**12. Supported UI/Unit test env for modules**

**13. Supported flavors configuration**

**Comment by michaem** [comment]
Future changes:

- android is "god module"
- reduce module types count

**Reply by stepango** [reply]
My current thinking is not to reduce module count but make a single target definition produce multiple gradle projects e.g.
Feature module could be structured in the current way
```
- feature-a
  - feature-a.build.gradle.kts
  - src
    - main
      - java
        - com.example.feature-a
          - api -- only included in source set of feature-a-api target
          - impl -- only uncluded in source set of feature-a-impl target
```
